### DISCUSSION TITLE
High CPU usage per tab in Ghostty on Manjaro Linux

### DISCUSSION THREAD

**Discussion by plinkr** [ghostty-org/ghostty#3658] [Feature Requests, Ideas] [answered]
I compiled the current version of Ghostty with `zig build -p $HOME/.local -Doptimize=ReleaseFast`
```sh
Ghostty 1.0.1-dev+0000000

Version
  - version: 1.0.1-dev+0000000
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.14.5
    runtime    : 4.14.5
  - libadwaita : enabled
    build      : 1.5.3
    runtime    : 1.5.3
```
The problem I am presenting is that Ghostty is using 100% of each processor with every tab open, with each new tab, the CPU usage spikes to 100% for that additional core.
In the screenshots I am attaching, you can see what I am explaining, each new tab shows an increase in CPU usage. That reddish color indicates background usage, and most apps that monitor CPU usage, like `htop` or `btop`, do not detect this type of background usage. That plugin is the `System monitor applet` from Cinnamon.

I am using Manjaro Linux, and I am attaching the relevant info from `sudo inxi -Fazy` here: 
```sh
System:
  Kernel: 5.15.165-1-MANJARO arch: x86_64 bits: 64 compiler: gcc v: 14.2.1
    clocksource: tsc avail: hpet,acpi_pm
    parameters: BOOT_IMAGE=/@/boot/vmlinuz-5.15-x86_64
    root=UUID=a9cbf476-db13-43bc-a095-49aa2591e826 rw rootflags=subvol=@ quiet
    udev.log_priority=3
  Desktop: Cinnamon v: 6.2.9 tk: GTK v: 3.24.43 wm: Muffin v: 6.2.0 tools:
    avail: cinnamon-screensaver dm: LightDM v: 1.32.0 Distro: Manjaro
    base: Arch Linux
CPU:
  Info: model: Intel Core i7-6700HQ socket: U3E1 bits: 64 type: MT MCP
    arch: Skylake-S gen: core 6 level: v3 note: check built: 2015
    process: Intel 14nm family: 6 model-id: 0x5E (94) stepping: 3
    microcode: 0xF0
  Topology: cpus: 1x dies: 1 cores: 4 threads: 8 tpc: 2 smt: enabled cache:
    L1: 256 KiB desc: d-4x32 KiB; i-4x32 KiB L2: 1024 KiB desc: 4x256 KiB
    L3: 6 MiB desc: 1x6 MiB
  Speed (MHz): avg: 2705 min/max: 800/3500 base/boost: 3100/8300 scaling:
    driver: intel_pstate governor: powersave volts: 1.1 V ext-clock: 100 MHz
    cores: 1: 2705 2: 2705 3: 2705 4: 2705 5: 2705 6: 2705 7: 2705 8: 2705
    bogomips: 41621
  Flags: avx avx2 ht lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx
  Vulnerabilities:
  Type: gather_data_sampling status: Vulnerable: No microcode
  Type: itlb_multihit status: KVM: VMX disabled
  Type: l1tf mitigation: PTE Inversion; VMX: conditional cache flushes, SMT
    vulnerable
  Type: mds mitigation: Clear CPU buffers; SMT vulnerable
  Type: meltdown mitigation: PTI
  Type: mmio_stale_data mitigation: Clear CPU buffers; SMT vulnerable
  Type: reg_file_data_sampling status: Not affected
  Type: retbleed mitigation: IBRS
  Type: spec_rstack_overflow status: Not affected
  Type: spec_store_bypass mitigation: Speculative Store Bypass disabled via
    prctl and seccomp
  Type: spectre_v1 mitigation: usercopy/swapgs barriers and __user pointer
    sanitization
  Type: spectre_v2 mitigation: IBRS; IBPB: conditional; STIBP: conditional;
    RSB filling; PBRSB-eIBRS: Not affected; BHI: Not affected
  Type: srbds mitigation: Microcode
  Type: tsx_async_abort mitigation: TSX disabled
Graphics:
  Device-1: Intel HD Graphics 530 vendor: CLEVO/KAPOK driver: i915 v: kernel
    arch: Gen-9 process: Intel 14n built: 2015-16 ports: active: eDP-1
    empty: none bus-ID: 00:02.0 chip-ID: 8086:191b class-ID: 0300
  Device-2: NVIDIA GM107M [GeForce GTX 960M] vendor: CLEVO/KAPOK
    driver: nvidia v: 550.107.02 alternate: nouveau,nvidia_drm non-free: 550.xx+
    status: current (as of 2024-09; EOL~2026-12-xx) arch: Maxwell code: GMxxx
    process: TSMC 28nm built: 2014-2019 pcie: gen: 2 speed: 5 GT/s lanes: 8
    link-max: gen: 3 speed: 8 GT/s lanes: 16 ports: active: none
    empty: DP-1,DP-2,HDMI-A-1 bus-ID: 01:00.0 chip-ID: 10de:139b class-ID: 0300
  Display: x11 server: X.org v: 1.21.1.13 with: Xwayland v: 24.1.2 driver: X:
    loaded: modesetting,nvidia gpu: i915 display-ID: :0 screens: 1
  Screen-1: 0 s-res: 1920x1080 s-size: <missing: xdpyinfo>
  Monitor-1: eDP-1 mapped: eDP-1-1 model: LG Display 0x046f built: 2014
    res: 1920x1080 hz: 60 dpi: 142 gamma: 1.2 size: 344x194mm (13.54x7.64")
    diag: 395mm (15.5") ratio: 16:9 modes: 1920x1080
  API: EGL v: 1.5 hw: drv: intel iris drv: nvidia platforms: device: 0
    drv: nvidia device: 2 drv: iris device: 3 drv: swrast gbm: drv: nvidia
    surfaceless: drv: nvidia x11: drv: nvidia inactive: wayland,device-1
  API: OpenGL v: 4.6.0 compat-v: 4.5 vendor: nvidia mesa v: 550.107.02
    glx-v: 1.4 direct-render: yes renderer: NVIDIA GeForce GTX 960M/PCIe/SSE2
    memory: 1.95 GiB
  API: Vulkan v: 1.3.279 layers: 1 device: 0 type: discrete-gpu
    name: NVIDIA GeForce GTX 960M driver: nvidia v: 550.107.02
    device-ID: 10de:139b surfaces: xcb,xlib device: 1 type: integrated-gpu
    name: Intel HD Graphics 530 (SKL GT2) driver: mesa intel v: 24.1.6-arch1.1
    device-ID: 8086:191b surfaces: xcb,xlib
Swap:
  Kernel: swappiness: 60 (default) cache-pressure: 100 (default) zswap: yes
    compressor: zstd max-pool: 20%
  ID-1: swap-1 type: file size: 512 MiB used: 488.4 MiB (95.4%) priority: -2
    file: /swap/swapfile
Info:
  Memory: total: 32 GiB available: 31.16 GiB used: 11.78 GiB (37.8%)
    igpu: 128 MiB
  Processes: 457 Power: uptime: 114d 12h 35m states: freeze,mem,disk
    suspend: deep avail: s2idle wakeups: 258 fails: 2 hibernate: platform
    avail: shutdown, reboot, suspend, test_resume image: 12.46 GiB
    services: csd-power,upowerd Init: systemd v: 256 default: graphical
    tool: systemctl
```
Here are the screenshots:
![Screenshot from 2024-12-28 00-15-36](https://github.com/user-attachments/assets/d17fb621-50fd-4263-a6ed-9e24502a7293)
![Screenshot from 2024-12-28 00-15-46](https://github.com/user-attachments/assets/db9a28c1-d4ab-4ba0-a609-c73e09b5d480)
![Screenshot from 2024-12-28 00-16-19](https://github.com/user-attachments/assets/2cd5353f-24c8-435b-a9fb-35889ebf2b20)
![Screenshot from 2024-12-28 00-16-29](https://github.com/user-attachments/assets/e0f6b915-3cb2-4c39-b03b-ea4650ebb56e)
![Screenshot from 2024-12-28 00-16-43](https://github.com/user-attachments/assets/f366dd3a-b338-4af6-839e-6815ea5e3a42)

**Comment by rhodes-b** [comment]
I dont think tabs are supposed to get too many unique resources but can you try setting
`gtk-single-instance = true` in your config
https://ghostty.org/docs/help/gtk-single-instance
has some more info on what it does

**Reply by 00-kat** [reply]
That's probably irrelevant.

**Reply by plinkr** [reply]
I changed the setting as you recommended, but the behavior is still the same

**Comment by 00-kat** [comment, marked as answer]
Maybe it's related to #3224.

**Reply by plinkr** [reply]
They may be related, the slowness described in that discussion is also something I see here on Manjaro

**Comment by philippehenin** [comment]
Just wanted to help. I have been using ghostty on different linux laptops for a few days. It is nice and works flawlessly.

On my chromebook (debian 12), I have the high CPU problem. I have tried the .deb install the v1.0.1 and v1.0.0 build from sources. I see the strange behavior. 

I did a small screencast to illustrate https://screencast.apps.chrome/1-MjA6K6nhMgpyFAp0fFtp8G53kCfBO3g?createdTime=2025-01-03T12%3A51%3A08.447Z

I hope it helps. It looks like software rendering vs. hardware video rendering, still searching…

I applied the "Epol"  proposal, it is way better. But I am still a bit concern. For me, it is still doing too much cpu. But it is useable now.

https://screencast.apps.chrome/1-cd5ZeBq7WBnYo4rv12abaG-3rLmchfl?createdTime=2025-01-03T13%3A48%3A30.253Z

Compare to a "native" debian 12 (on old slower laptop) with normal cpu usage: https://youtu.be/i0-Jh3BXbMg.

P.

**Reply by plinkr** [reply]
In the second screencast you show, the ghostty you have in your `Gits/ghostty` folder is indeed compiled with the fixes using `epoll` instead of `io_uring`, but you are running the `ghostty` that is in your $PATH, which still has the backend using `io_uring`, and that's why it continues to show elevated iowait usage. Use the `ghostty` compiled with the changes, located at: `~/Gits/ghostty/zig-out/bin/ghostty`. 

In the screencast at the 0:40 second mark when starting `ghostty`, you can see that the backend is still set to `io_uring` and not `epoll` as it should be with the fix. 

When starting `ghostty`, the console should display this:

```
$ zig-out/bin/ghostty 
info: ghostty version=1.0.2-dev+0000000
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.epoll
```
take special notice to the line `info: libxev backend=main.Backend.epoll`

**Reply by philippehenin** [reply]
Effectively, my bad for the command.

No problem anymore. CPU normal with the correct executable.

```
philippehenin@penguin:~$ Gits/ghostty/zig-out/bin/ghostty 
info: ghostty version=1.0.1
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
**info: libxev backend=main.Backend.epoll**
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.8.3 runtime=4.8.3
info: reading configuration file path=/home/philippehenin/.config/ghostty/config
info(config): default shell source=env value=/bin/bash
info(sentry): sentry envelope does not contain crash, discarding
info(gtk): libadwaita version build=1.2.2 runtime=1.2.2
```

Thank you for the great reply and the great terminal.
P.

**Comment by gzhechu** [comment]
I meet same situation in my openSUSE slowroll. I am using opensuse-slowroll with most recent update, KDE destop environment. I installed ghostty from prebuild binary for openSUSE following this instruction: https://ghostty.org/docs/install/binary#opensuse, and I noticed when I start ghostty, the CPU monitor applet shows servel high CPU usage, when I add a new tab, the CPU bar in applet synchronization increased. After I quit ghostty, the CPU bar resumed to normal.

![屏幕截图_20250117_170355](https://github.com/user-attachments/assets/78fa186a-577b-44e2-999c-b60a3cbb7263)


> 
> robin@k8plus:~> cat /etc/os-release 
> NAME="openSUSE Tumbleweed-Slowroll"
> # VERSION="20250101"
> ID="opensuse-slowroll"
> ID_LIKE="opensuse-tumbleweed opensuse suse"
> VERSION_ID="20250101"
> PRETTY_NAME="openSUSE Tumbleweed-Slowroll"
> ANSI_COLOR="0;32"
> # CPE 2.3 format, boo#1217921
> CPE_NAME="cpe:2.3:o:opensuse:slowroll:20250101:*:*:*:*:*:*:*"
> #CPE 2.2 format
> #CPE_NAME="cpe:/o:opensuse:slowroll:20250101"
> BUG_REPORT_URL="https://bugzilla.opensuse.org"
> SUPPORT_URL="https://bugs.opensuse.org"
> HOME_URL="https://www.opensuse.org"
> DOCUMENTATION_URL="https://en.opensuse.org/openSUSE:Slowroll"
> LOGO="distributor-logo-Slowroll"
> 
> 
> robin@k8plus:~> ghostty --version
> Ghostty 1.0.1
> 
> Version
>   - version: 1.0.1
>   - channel: stable
> Build Config
>   - Zig version: 0.13.0
>   - build mode : builtin.OptimizeMode.ReleaseFast
>   - app runtime: apprt.Runtime.gtk
>   - font engine: font.main.Backend.fontconfig_freetype
>   - renderer   : renderer.OpenGL
>   - libxev     : main.Backend.io_uring
>   - GTK version:
>     build      : 4.16.12
>     runtime    : 4.16.7
>   - libadwaita : enabled
>     build      : 1.6.3
>     runtime    : 1.6.2
>   - libX11     : enabled
> robin@k8plus:~>

**Reply by plinkr** [reply]
Please read this post: https://github.com/ghostty-org/ghostty/discussions/3224. It describes the reason for this problem, and one solution is to recompile ghostty by following the steps in this comment: https://github.com/ghostty-org/ghostty/discussions/3224#discussioncomment-11691632
